### DISCUSSION TITLE
How to setup Dictation with Whisper

### DISCUSSION THREAD

**Discussion by santiyounger** [SemlerPDX/OpenAI-VoiceAttack-Plugin#6] [Q&A] [answered]
Hey loving your work, than you so much for sharing this project!

I installed the plugin and installed the AVCS profile too.

Everything is working fine, however as a beginner even after reading the manual I can't figure out how to create a new command to dictate using Open AI's whisper.

Ideally I was hoping there was a pre-made command to achieve this, however I couldn't find one in the ones you created, however I could be wrong.

I see how it makes sense that there's a way to record the audio and send that to whisper and then it deleted the audio, but I'm lost on how to actually set this up step-by-step as a beginner.

Appreciate any help, excited to start using it.

Thank you!

**Comment by SemlerPDX** [comment]
One of the challenging things I programmed into this plugin was a way for the ChatGPT plugin context to offer a GetInput phase which can take advantage of an advanced dictation-until-silence method and stitch together any audio files produced in order to simplify the subsequent Whisper API call to a single audio file transcription/translation.  Because my code base is open source, you could create an inline function in C# inside a VoiceAttack command where you could try to adapt or approximate my methods of accomplishing my goal - attribution and reference to source would be appreciated in the comments.  These methods are in the service class "Dictation.cs"

**Comment by SemlerPDX** [comment]
The way this class is called is from within the ChatGPT context code starting at line 541:
```csharp
// Get User Input from variable, dictation token, or begin listening for dictation:
userInput = Dictation.GetUserInput(useWhisper, sayPreListen);

// If processing dictation audio with OpenAI Whisper transcription/translation:
if (useWhisper && !String.IsNullOrEmpty(userInput) && userInput != DEFAULT_REPROCESS_FLAG)
	userInput = Whisper.GetUserInput(operation)
```

**Comment by Robertsmania** [comment]
I think the OP is looking for a way to setup commands within VA rather than getting that deep into the plugin code?

In my adaptation, I have a plugin context "Transcribe_Audio" that handles getting the dictation recording, sending it to Whisper and setting a ~~Transcription_Response VA text variable with the response.

```
Say, 'What would you like to say?'  (and wait until it completes)
Execute command, 'A2 TwtichChat -Transcribe Audio' (and wait until it completes)
Begin Text Compare : [~~Transcription_Response] Does Not Equal ''
    Execute command, 'A1 - SpeechCoordinator - Send_iRacingChatMessage' (and wait until it completes)
Else
    Say, 'I didn't get that.'  (and wait until it completes)
    Write [Red] 'No dictation transcription?' to log
End Condition
```
Does your GetInput phase have something similar they can use?

**Comment by santiyounger** [comment]
Thank you @SemlerPDX  and @Robertsmania 

Appreciate the help! And yes @Robertsmania is right, unfortunately I'm a complete beginner when it comes to programming and code, so I'm quite lost on how to simply use whisper dictation for my own use, I don't want to alter your method or create my own code, I just want to use the dictation method with Whisper.

@Robertsmania your response was helpful 🙏 but I still don't fully know how to  set that up in VA correctly

Just for clarity @SemlerPDX. All I want is a way to dictate in Voiceattack using Whisper so that I can have a transcript I can insert into a writing app.

I couldn't find any pre-made command that did this in the plugin nor in the AVCS profile.
Perhaps there is one I just couldn't find it.

Would it be possible to have a command I can download/import and use as a normal user?

Ideally installing a command in VA for simple use would be great, but if that's hard on your side. Appreciate instructions to do it myself assuming I'm a 5 year old non-programmer hahaha

Thank you @SemlerPDX and @Robertsmania

**Comment by SemlerPDX** [comment]
> Thank you @SemlerPDX and @Robertsmania
> 
> Appreciate the help! And yes @Robertsmania is right, unfortunately I'm a complete beginner when it comes to programming and code, so I'm quite lost on how to simply use whisper dictation for my own use, I don't want to alter your method or create my own code, I just want to use the dictation method with Whisper.
> 
> @Robertsmania your response was helpful 🙏 but I still don't fully know how to set that up in VA correctly
> 
> Just for clarity @SemlerPDX. All I want is a way to dictate in Voiceattack using Whisper so that I can have a transcript I can insert into a writing app.
> 
> I couldn't find any pre-made command that did this in the plugin nor in the AVCS profile. Perhaps there is one I just couldn't find it.
> 
> Would it be possible to have a command I can download/import and use as a normal user?
> 
> Ideally installing a command in VA for simple use would be great, but if that's hard on your side. Appreciate instructions to do it myself assuming I'm a 5 year old non-programmer hahaha
> 
> Thank you @SemlerPDX and @Robertsmania

No worries, if this sort of thing is a bit too technical, I can help.  I see what you mean now, and what you want this for.  I can probably whip up some sort of command system that will work for you, I'll just modularize my Dictation class in this plugin into an inline function for you as a more advanced 'dictation until silence' which you can use in a writing app.

Won't need instructions beyond knowing how to call the command, and that when it is finished, a bunch of key presses will occur, or a 'paste' command, not sure yet but I'll write it up and see how it turns out.

**Comment by santiyounger** [comment]
Thank you, you are the best, a command I can import into VoiceAttack would be truly appreciated 🙏

If you ever need exposure with your projects on my YouTube channel or Twitter let me know

I recorded a short video for you with some examples on how I've though of this command to work, in case it's helpful on your side

[quick video on VoiceAttack AI Plugin and Whisper | Drobpox Video
](https://www.dropbox.com/scl/fi/zgxmi564bgkea8k7lsvcc/quick-video-on-VoiceAttack-AI-Plugin-and-Whisper.mp4?rlkey=oahbso1xe3osq033qtlbw313d&dl=0)

Thank you!

**Comment by Robertsmania** [comment]
Exposure is always good!  Here's an example of how I'm using the Whisper dictation to isolate viewers in my Twitch chat:
https://www.youtube.com/watch?v=4Q3ydrvN8tI
And another (slightly clumsy) example when I was still just getting it to work but shows a practical application of banning a spam bot that came into my channel offering to sell followers:
https://www.youtube.com/watch?v=hfiRjBldgAI

Its always been hard since the screen names are usually spelled with weird characters and sequences that do not match typical phonetic patterns.  And in the past, the native dictation accuracy for anything was so poor, it was pretty much a lost cause.

Now with the much higher accuracy of the Whisper model and the ability to pass in a prompt that includes all the unusual spellings of screen names and such, I can actually get very reliable results.

In the use case above its actually a two step process.  I send the audio from my dictation to Whisper with a prompt that includes all the screen names and the last several messages from the twitch chat.  That returns text with what it thinks I said, and if it matched someone's screen name it usually has it spelled the way they have it.

Then I build a prompt to send off to ChatGPT including the dictation response as the 'search term' and give it similar information about the current users in chat and recent messages for context.  It looks at that data and responds with who it thinks is the most likely user.
```
  {
    "role": "user",
    "content": "Help me identify a Twitch chat viewer from this search term: \"some form of greetings\". The search term may be similar to a username or reference something that was said in chat. Please help idetify the most appropriate username. If its unclear please say so and do not make up an answer. Keep your reply very brief, no need to explain how you got your result or repeat the search term. All the current usernames: Robertsmaniac. Recent Twitch Chat messages formatted as (username: content): (Robertsmaniac: greetings)"
  }
```

Then I take whatever ChatGPT responds and search the list of actual viewers in chat to see if there is a match.

The truth is I actually check just the whisper response first and the accuracy is so good I nearly always find a match if the search term was reasonable.  But I send it off to ChatGPT anyway just for the extra flavor and comic relief.  And indeed, sometimes the AI is able to find the right user even when the direct text comparison did not.

Once you have reliable and accurate dictation, lots of opportunities open up!

**Comment by SemlerPDX** [comment, marked as answer]
> Thank you, you are the best, a command I can import into VoiceAttack would be truly appreciated 🙏
> 
> If you ever need exposure with your projects on my YouTube channel or Twitter let me know
> 
> I recorded a short video for you with some examples on how I've though of this command to work, in case it's helpful on your side
> 
> [quick video on VoiceAttack AI Plugin and Whisper | Drobpox Video ](https://www.dropbox.com/scl/fi/zgxmi564bgkea8k7lsvcc/quick-video-on-VoiceAttack-AI-Plugin-and-Whisper.mp4?rlkey=oahbso1xe3osq033qtlbw313d&dl=0)
> 
> Thank you!

Okay, I think I understand your goal better now.  If you just need a Dictation Mode typing assistant which uses more accurate transcription than Windows Speech via the OpenAI plugin and the Whisper context, it can all be done with standard command actions in VoiceAttack and calling the OpenAI plugin as normal using the `Whisper.Transcribe` context.

I've tossed together an example profile you can use, and/or modify as needed to better suit your individual needs.  It works by having the user say a starting voice command, such as, "Start Dictation" - while you speak, your dictation creates tiny audio files which are sent to Whisper for transcription, doesn't matter if you pause in your speech or use one long run on sentence followed by more pauses and more speech.  It will keep gathering (in order) the things you say, send them off to Whisper for transcription, all in the background while Dictation is enabled - when you are done, you'd use another voice command to end it, such as, "End Dictation" ... final processing will occur (if not already done), and then will paste the combined transcriptions to the cursor location.

Edit the two actual voice commands as needed, avoid altering function commands unless you are familiar with what they are doing and understand how to modify them.  If you have any questions, or find a bug that I didn't during my testing this afternoon, let me know here and I'll help out.

#### Download Here:
<a href="https://veterans-gaming.com/semlerpdx/avcs/dictate/">OpenAI Whisper Dictation Example-(v1.0)-Profile.vap</a>

**Comment by SemlerPDX** [comment]
_**I will most likely make this example profile above into a proper AVCS public profile, but just wanted to toss together the concept and get it to you early so maybe you could provide feedback.  If it's good to go, it will become AVCS TYPE v1.0 soon.**_

**Comment by SemlerPDX** [comment]
Unfortunately, this evening (same day as I first posted the link above to this example profile), the OpenAI Whisper API has gone down.  This is on the OpenAI end, and likely some form of technical difficulties. Just a terrible coincidence to happen this evening, but they've been fairly quick to restore services in past that experienced (rare) issues.  This is a new one to me, but I'm not gonna worry so long as it comes back by Thursday/Friday of this week at the latest.  Keeping an eye on the 'whisper' channel at the OpenAI Discord until then.

If you experience errors or simply no output from this example profile (near the date/time of this reply), just give it a day, and try again.

Cheers!
-Sem

**Comment by santiyounger** [comment]
Thanks a lot @SemlerPDX. This is incredible, truly appreciate it 🙏
 Like you said in discord, seems like Whisper is back, thanks for the update.

I've installed these commands in Voice Attack, and it's great!

  Here's some feedback that I hope will be helpful:

## line breaks 

I saw that the command mentions line breaks in pauses, by default line breaks aren't working, I've tried to insert my dictation in various apps including notepad, and google docs, but there's no line break, I like this concept, it'd be pretty helpful.

## double spaces

by default (without modifying anything in he command) I seem to get double spaces and no line breaks.

### text I read for dictation test 

I got some text from CHATGPT to test dictation, here's the original text I read for the test below it.
```

The moon landing on July 20, 1969, is a landmark in human history. 

NASA's Apollo 11 mission made it possible.

Neil Armstrong and Buzz Aldrin walked on the moon.

Michael Collins orbited in the command module.

Armstrong said, "That's one small step for man, one giant leap for mankind."

This event marked the peak of the U.S.-Soviet space race during the Cold War.
```

## results of dictation test

The results were no line breaks with  double spaces, after pauses, that I have to manually remove.

dicatation test:

  the moon landing on July 20, 1969.  is a landmark in human history. ` `NASA's Apollo 11 mission made it possible.`  `Neil Armstrong and Buzz Aldrin walked on the moon.`  `michael collins orbiting the command module  Armstrong said, that's one small step for man, one giant leap for mankind. ` `This event marked the peak of the US-Soviet space race during the Cold War.


## dictation sometimes incomplete

 I'm actually dictating this text. (although I've manually corrected it). And what I've noticed is that it takes me a few attempts because some segments of the dictation are sometimes incomplete in the transcript.

I  am using a high quality mic very close to my mouth.  So I don't know why this tends to happen.

 I've tested making long pauses (10 secs) and short ones (2 secs) but I can't tell if these incomplete transcripts are causes by pauses or not.


Thanks, really enjoying it, thank you for putting this together!

**Reply by SemlerPDX** [reply]
Thank you for testing this out!  The options for separators which I added is a very basic example.  I added a global text variable which can hold the contents of what should separate individual dictation transcriptions (between pauses in speech).  I included an option example for a double line break, and a double space break:
![Example actions](https://i.imgur.com/mNeFLC1.png)

You would need to disable the one you don't want, and leave the one you do want enabled.  You can also alter these as needed for whatever separator you want - again, this is only an example.  There are many ways this could be improved, including post- or pre-processing of the dictation transcription products.  Currently, the example profile is simply building a string of strings.

I have not experienced any loss or incomplete transcriptions, so I'll test and play around with it today to see if I can replicate that.  There shouldn't technically be any loss, each dictation audio file is enqueued as a command to send to Whisper API (in order) and the products retrieved are added to the string we build which contains all the transcriptions in the order in which they were enqueued.

Now that the Whisper API is back online, I'm wondering if there are any issues.  It may be wise to check your OpenAI plugin error log from time to time.

You can find this file here:
`%AppData%\OpenAI_VoiceAttack_Plugin\openai_errors.log`

**Reply by SemlerPDX** [reply]
**To be clear about what I mean by pre- or post-processing:**

Imagine if you want to dictate an entire letter, and your line separator option is set to double-space.  You might say a few dictations to comprise the first paragraph, and then perhaps say something by itself (after a speech pause) like "Paragraph break", and the another speech pause to ensure it is nice an isolated.

Then, inside the 'Stop Dictation' command (before using the product), you could post-process this text variable to check for that phrase (followed by the default separator) to cull it from the transcription and replace it with something like three new lines `{NEWLINE}{NEWLINE}{NEWLINE}` like this:
![Example image 2](https://i.imgur.com/VxJ0rsT.png)

Such a system would require testing and slight modifications (sometimes even accounting for curve-balls like, "what if it transcribes brake instead of break?"), but is a good example on how depth and complexity could be added to transcription.

**Reply by santiyounger** [reply]
thank you @SemlerPDX that makes sense! I just tested what you said about disabling one or the other for `new line` vs `space` that seems to solve the issue. 

To solve the double space I replaced `{SPACE}{SPACE}` for just one `{SPACE}` that seems to remove the double space for me.


## idea for better UX

 I personally like sound effects like the one you have in your other command `start chat` 

so I replaced the TTS voice that says `"dictation started"` and  `"dictation ended"` for some cool windows sound effect.

The main reason why I think this works better, is because I use a speaker on my PC so at the start of every transcript Whisper included the words `"dictation started" ` at the beginning of the text, this is why I think having a sound effect is a better approach, for better user experience

Thanks for the tip on error logs, I did have a few errors, here's the file in case that helps with trying to replicate the incomplete transcript issues.
[openai_errors.log](https://github.com/SemlerPDX/OpenAI-VoiceAttack-Plugin/files/13056585/openai_errors.log)


Thank you!

**Reply by SemlerPDX** [reply]
That makes perfect sense to not use text-to-speech for UX indicators if you're not using headphones.  >90% of VoiceAttack users use headphones, but sound effects or even visual cues are perfectly sufficient for those users as an alternative.

Another option for such use cases is to use a push-to-talk system which can unmute listening in VoiceAttack while it is held down, though this might not be ideal for all users who do not prefer using headphones.

If you want to address the actual sounds I use in my AVCS CHAT profile, you can find the audio files in the VoiceAttack programs folder under the `Sounds\OpenAI_Plugin` folder.  They're just copies from Windows anyway, and can also be found in the folder containing Windows sounds.

Best wishes and good luck!  Let me know if you have any further issues, or questions.  I'll test today to see if I can replicate transcriptions being lost or not present in a long dictation session - hopefully it's something simple or obvious (and on MY end where I can fix it rather than some OpenAI API quirk which I cannot fix).

Cheers!

**Reply by santiyounger** [reply]
Perfect, great instructions, I managed to add the sound effect for my case! I might experiment with activation through shortcuts too.

On your previous comment you shared a really great idea on how to activate line breaks through speech, that's incredible. It might be a bit too complicated for me to integrate that system, but if that's something you can add in the future, please let me know, I'd love to test it and provide feedback.

Thank you, if you see any issues on your side please keep me updated, otherwise thanks for all your incredible work, super happy with the results!

**Reply by SemlerPDX** [reply]
On second thought, it may be as simple as creating another voice command literally with the phrase(s) to add a paragraph break, and then enqueue a command which will do so - this should work because recognized voice commands are naturally excluded from dictation:
![Image Example 1](https://i.imgur.com/EuQ2avL.png)
...and the command which gets enqueued simply adds some newlines to the end of the existing string of text being built:
![Image Example 2](https://i.imgur.com/3iVbIBI.png)

**Reply by santiyounger** [reply]
Interesting approach, I'll try to  learn VoiceAttack better to figure out cool setups like this, thank you!
